flang/lib/Lower/CallInterface.cpp

```diff
@@ -0,0 +1,455 @@
+}
+
+mlir::Location Fortran::lower::CallerInterface::getCalleeLocation() const {
+  const auto &proc = procRef.proc();
+  if (const auto *symbol = proc.GetSymbol()) {
+    // FIXME: If the callee is defined in the same file but after the current
+    // unit we do cannot get its location here and the funcOp is created at the
+    // wrong location (i.e, the caller location).
+    if (const auto *details =
+            symbol->detailsIf<Fortran::semantics::ProcEntityDetails>())
+      if (const auto *interfaceSymbol = details->interface().symbol())
+        symbol = interfaceSymbol;
+    return converter.genLocation(symbol->name());
+  }
+  // Unknown location for intrinsics.
+  return converter.genLocation();
+}
+
+Fortran::evaluate::characteristics::Procedure
+Fortran::lower::CallerInterface::characterize() const {
+  auto &foldingContext = converter.getFoldingContext();
+  auto characteristic =
+      Fortran::evaluate::characteristics::Procedure::Characterize(
+          procRef.proc(), foldingContext.intrinsics());
+  assert(characteristic && "Failed to get characteristic from procRef");
+  // The characteristic may not contain the argument characteristic if no
+  // the ProcedureDesignator has no interface.
+  if (!characteristic->HasExplicitInterface()) {
+    for (const auto &arg : procRef.arguments()) {
+      // Argument cannot be optional with implicit interface
+      const auto *expr = arg.value().UnwrapExpr();
+      assert(expr &&
+             "argument in call with implicit interface cannot be assumed type");
+      auto argCharacteristic =
+          Fortran::evaluate::characteristics::DummyArgument::FromActual(
+              "actual", *expr, foldingContext);
+      assert(argCharacteristic &&
+             "failed to characterize argument in implicit call");
+      characteristic->dummyArguments.emplace_back(
+          std::move(*argCharacteristic));
+    }
+  }
+  return *characteristic;
+}
+
+void Fortran::lower::CallerInterface::placeInput(
+    const PassedEntity &passedEntity, mlir::Value arg) {
+  assert(static_cast<int>(actualInputs.size()) > passedEntity.firArgument &&
+         passedEntity.firArgument >= 0 &&
+         passedEntity.passBy != CallInterface::PassEntityBy::AddressAndLength &&
+         "bad arg position");
+  actualInputs[passedEntity.firArgument] = arg;
+}
+
+void Fortran::lower::CallerInterface::placeAddressAndLengthInput(
+    const PassedEntity &passedEntity, mlir::Value addr, mlir::Value len) {
+  assert(static_cast<int>(actualInputs.size()) > passedEntity.firArgument &&
+         static_cast<int>(actualInputs.size()) > passedEntity.firLength &&
+         passedEntity.firArgument >= 0 && passedEntity.firLength >= 0 &&
+         passedEntity.passBy == CallInterface::PassEntityBy::AddressAndLength &&
+         "bad arg position");
+  actualInputs[passedEntity.firArgument] = addr;
+  actualInputs[passedEntity.firLength] = len;
+}
+
+bool Fortran::lower::CallerInterface::verifyActualInputs() const {
```

Comment:
Shouldn't the arguments be verified by semantics?

Reply:
Yes, its doing it, this function is only verifying that the user has placed all the lowered arguments (with `placeInput`) before calling `getActualInputs()`.